pkg/archive/compress.go

```diff
@@ -22,41 +23,91 @@ const (
 	CompressGzip
 	// CompressXz compression
 	CompressXz
+	// CompressZstd compression
+	CompressZstd
 )
 
 // compressHeaders are used to detect the compression type
 var compressHeaders = map[CompressType][]byte{
 	CompressBzip2: []byte("\x42\x5A\x68"),
 	CompressGzip:  []byte("\x1F\x8B\x08"),
 	CompressXz:    []byte("\xFD\x37\x7A\x58\x5A\x00"),
+	CompressZstd:  []byte("\x28\xB5\x2F\xFD"),
 }
 
 func Compress(r io.Reader, oComp CompressType) (io.Reader, error) {
 	br := bufio.NewReader(r)
-	head, err := br.Peek(10)
+
+	var rComp CompressType
+
+	// Peek length of the magic header for the given compression into the stream
+	head, err := br.Peek(len(compressHeaders[oComp]))
 	if err != nil {
-		return br, err
+		if err == io.EOF {
+			// Not enough bytes to peek, that usually means the stream is shorter than
+			// length of the magic bytes header which is a sign of uncompressed stream.
+			rComp = CompressNone
+		} else {
+			return br, err
+		}
+	} else {
+		// Detect the compression type
+		rComp = DetectCompression(head)
 	}
-	rComp := DetectCompression(head)
+
+	// If the detected compression matches the requested compression, then return.
 	if rComp == oComp {
 		return br, nil
 	}
+
+	var dbr io.Reader = br
+	// If the stream is already compressed, then decompress it first
+	if oComp != CompressNone {
+		dbr, err = Decompress(br)
+		if err != nil {
+			println("here")
+			return nil, err
+		}
+	}
+
 	switch oComp {
 	case CompressGzip:
-		switch rComp {
-		case CompressNone:
-			return compressGzip(br)
-		case CompressBzip2:
-			return compressGzip(bzip2.NewReader(br))
-		case CompressXz:
-			cbr, _ := xz.NewReader(br)
-			return compressGzip(cbr)
-		}
+		return compressGzip(dbr)
```

Comment:
Before this change this switch case was a matrix of decompression and compression types where only gzip supported as the only output, now it supports all of the compression types as both decompression (input) and compression (output) with the exception of bzip2 as the output because of https://github.com/golang/go/issues/4828.